src/relations/mysql_tls.py

```diff
@@ -0,0 +1,237 @@
+
+    # =======================
+    #  Event Handlers
+    # =======================
+    def _on_set_tls_private_key(self, event: ActionEvent) -> None:
+        """Action for setting a TLS private key."""
+        self._request_certificate(event.params.get("internal-key", None))
+
+    def _on_tls_relation_joined(self, _) -> None:
+        """Request certificate when TLS relation joined."""
+        self._request_certificate(None)
+
+    def _on_certificate_available(self, event: CertificateAvailableEvent) -> None:
+        """Enable TLS when TLS certificate available."""
+        if (
+            event.certificate_signing_request.strip()
+            != self.charm.get_secret(SCOPE, "csr").strip()
+        ):
+            logger.error("An unknown certificate expiring.")
+            return
+
+        self.charm.unit.status = MaintenanceStatus("Setting up TLS")
+
+        self.charm.set_secret(
+            SCOPE, "chain", "\n".join(event.chain) if event.chain is not None else None
+        )
+        self.charm.set_secret(SCOPE, "cert", event.certificate)
+        self.charm.set_secret(SCOPE, "ca", event.ca)
+
+        self.push_tls_files_to_workload()
+        self.create_tls_config_file()
+
+        # set member-state to avoid unwanted health-check actions
+        self.charm.unit_peer_data["member-state"] = "waiting"
+        # trigger rolling restart
+        self.charm.on[self.charm.restart_manager.name].acquire_lock.emit()
+
+    def _on_certificate_expiring(self, event: CertificateExpiringEvent) -> None:
+        """Request the new certificate when old certificate is expiring."""
+        if event.certificate != self.charm.get_secret(SCOPE, "cert"):
+            logger.error("An unknown certificate expiring.")
+            return
+
+        key = self.charm.get_secret(SCOPE, "key").encode("utf-8")
+        old_csr = self.charm.get_secret(SCOPE, "csr").encode("utf-8")
+        new_csr = generate_csr(
+            private_key=key,
+            subject=self.charm.get_hostname_by_unit(self.charm.unit.name),
+            organization=self.charm.app.name,
+            sans=self._get_sans(),
+        )
+        self.certs.request_certificate_renewal(
+            old_certificate_signing_request=old_csr,
+            new_certificate_signing_request=new_csr,
+        )
+        self.charm.set_secret(SCOPE, "csr", new_csr.decode("utf-8"))
+
+    def _on_tls_relation_broken(self, _) -> None:
+        """Disable TLS when TLS relation broken."""
+        self.charm.set_secret(SCOPE, "ca", None)
+        self.charm.set_secret(SCOPE, "cert", None)
+        self.charm.set_secret(SCOPE, "chain", None)
+        self.remove_tls_config_file()
+        # set member-state to avoid unwanted health-check actions
+        self.charm.unit_peer_data["member-state"] = "waiting"
+        # trigger rolling restart
+        self.charm.on[self.charm.restart_manager.name].acquire_lock.emit()
```

Comment:
to confirm my understanding: does setting `communicationStack = MYSQL` just enable the ability to use TLS for XProtocol, but is not activated until the TLS CA, cert and chain are available? do we need to (or can we do) something to move cluster `communicationStack = XCOM`? (Does `communicationStack = MYSQL` fail if TLS files are not present?

Reply:
No. Using mysql as commstack enables encryption (and [other features](https://dev.mysql.com/doc/refman/8.0/en/group-replication-connection-security.html)) by default for the communication between group replication nodes.
It's independent of encryption with clients - so that's why we are enabling it no matter what.
Also, make the implementation simpler, because remove the necessity of having both internal and external keys.